craft_parts/executor/part_handler.py

```diff
@@ -73,34 +78,148 @@ def __call__(
+                    destdir=destdirs[dst_partition],
+                    sub_path=sub_path,
+                    dst_partition=dst_partition,
+                    permissions=permissions,
+                )
+        else:
+            # Ignore the layout and migrate from/to the same partition
+            self._migrate(
+                refdir=refdir,
+                srcdir=srcdir,
+                destdir=destdirs[self._src_partition],
+                sub_path="",
+                dst_partition=self._src_partition,
+                permissions=permissions,
+            )
+
+    def _migrate(
         self,
         refdir: Path,
         srcdir: Path,
         destdir: Path,
+        sub_path: str,
+        dst_partition: str,
         permissions: list[Permissions] | None = None,
     ) -> None:
-        visible_files, visible_dirs = overlays.visible_in_layer(refdir, destdir)
+        """Actually migrate content from a source to a destination.
+
+        Associate the lists of migrated content to the partition in a map to
+        later store it in the proper state.
+        """
+        visible_files, visible_dirs = overlays.visible_in_layer(
+            refdir / sub_path,
+            destdir,
+        )
+        logger.debug(f"excluding already migrated files: {self.all_migrated_files()}")
+        visible_files = visible_files - self.all_migrated_files()
+        logger.debug(
+            f"excluding already migrated dirs: {self.all_migrated_directories()}"
+        )
+        visible_dirs = visible_dirs - self.all_migrated_directories()
+
         layer_files, layer_dirs = migration.migrate_files(
             files=visible_files,
             dirs=visible_dirs,
-            srcdir=srcdir,
+            srcdir=srcdir / sub_path,
             destdir=destdir,
             oci_translation=True,
             permissions=permissions,
         )
-        self.migrated_files |= layer_files
-        self.migrated_dirs |= layer_dirs
+        if dst_partition not in self.migrated_files:
+            self.migrated_files[dst_partition] = {}
+
+        for f in layer_files:
+            src_path = str(Path(sub_path) / f)
+            self.migrated_files[dst_partition][src_path] = f
+
+        if dst_partition not in self.migrated_directories:
+            self.migrated_directories[dst_partition] = {}
 
-    def get_state(self) -> MigrationState:
-        return MigrationState(files=self.migrated_files, directories=self.migrated_dirs)
+        for f in layer_dirs:
+            src_path = str(Path(sub_path) / f)
+            self.migrated_directories[dst_partition][src_path] = f
+
+    def all_migrated_files(self) -> set[str]:
+        """Merge lists of files migrated to every partitions.
+
+        Return a list of paths relative to the source partition.
+        """
+        migrated_files: set[str] = set()
+        for m in self.migrated_files.values():
+            migrated_files |= set(m)
+        return migrated_files
```

Comment:
We can shorten this to `return set(self.migrated_files.values())`. Also this could be a property? And can `self.migrated_files` be private now that we have this?

Reply:
I do not think we can do what you suggest. `migrated_files` (and `migrated_directories`) are not of type `dict[str, _Migrated_Contents]`, so dicts of dicts. However the name may now be misleading. What do you think of ` partitions_migrated_files`?

> Also this could be a property?

Correct! I will fix that.

> And can self.migrated_files be private now that we have this?

No, because this is still used to call `consolidate_states()` and `_clean_dangling_whiteouts()` in `_migrate_overlay_files_to_stage()`